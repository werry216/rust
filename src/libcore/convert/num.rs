mod private {
    /// This trait being unreachable from outside the crate
    /// prevents other implementations of the `FloatToInt` trait,
    /// which allows potentially adding more trait methods after the trait is `#[stable]`.
    #[unstable(feature = "convert_float_to_int", issue = "67057")]
    pub trait Sealed {}
}

/// Supporting trait for inherent methods of `f32` and `f64` such as `round_unchecked_to`.
/// Typically doesnâ€™t need to be used directly.
#[unstable(feature = "convert_float_to_int", issue = "67057")]
pub trait FloatToInt<Int>: private::Sealed + Sized {
    #[cfg(not(bootstrap))]
    #[unstable(feature = "float_approx_unchecked_to", issue = "67058")]
    #[doc(hidden)]
    unsafe fn approx_unchecked(self) -> Int;
}

macro_rules! impl_float_to_int {
    ( $Float: ident => $( $Int: ident )+ ) => {
        #[unstable(feature = "convert_float_to_int", issue = "67057")]
        impl private::Sealed for $Float {}
        $(
            #[unstable(feature = "convert_float_to_int", issue = "67057")]
            impl FloatToInt<$Int> for $Float {
                #[cfg(not(bootstrap))]
                #[doc(hidden)]
                #[inline]
                unsafe fn approx_unchecked(self) -> $Int {
                    crate::intrinsics::float_to_int_approx_unchecked(self)
                }
            }
        )+
    }
}

impl_float_to_int!(f32 => u8 u16 u32 u64 u128 usize i8 i16 i32 i64 i128 isize);
impl_float_to_int!(f64 => u8 u16 u32 u64 u128 usize i8 i16 i32 i64 i128 isize);
