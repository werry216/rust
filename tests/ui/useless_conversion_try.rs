#![deny(clippy::useless_conversion)]

use std::convert::TryFrom;

fn test_generic<T: Copy>(val: T) -> T {
    T::try_from(val).unwrap()
}

fn test_generic2<T: Copy + Into<i32> + Into<U>, U: From<T>>(val: T) {
    let _ = U::try_from(val).unwrap();
}

fn main() {
    test_generic(10i32);
    test_generic2::<i32, i32>(10i32);

    let _: String = TryFrom::try_from("foo").unwrap();
    let _ = String::try_from("foo").unwrap();
    #[allow(clippy::useless_conversion)]
    let _ = String::try_from("foo").unwrap();

    let _: String = TryFrom::try_from("foo".to_string()).unwrap();
    let _ = String::try_from("foo".to_string()).unwrap();
    let _ = String::try_from(format!("A: {:04}", 123)).unwrap();
}
