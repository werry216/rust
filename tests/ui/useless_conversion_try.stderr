error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:6:13
   |
LL |     let _ = T::try_from(val).unwrap();
   |             ^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> $DIR/useless_conversion_try.rs:1:9
   |
LL | #![deny(clippy::useless_conversion)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider removing `T::try_from()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:7:5
   |
LL |     val.try_into().unwrap()
   |     ^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:29:21
   |
LL |     let _: String = "foo".to_string().try_into().unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:30:21
   |
LL |     let _: String = TryFrom::try_from("foo".to_string()).unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `TryFrom::try_from()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:31:13
   |
LL |     let _ = String::try_from("foo".to_string()).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `String::try_from()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:32:13
   |
LL |     let _ = String::try_from(format!("A: {:04}", 123)).unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `String::try_from()`

error: Useless conversion to the same type
  --> $DIR/useless_conversion_try.rs:33:21
   |
LL |     let _: String = format!("Hello {}", "world").try_into().unwrap();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider removing `.try_into()`

error: aborting due to 7 previous errors

